common/ast/src/parser/rule/expr.rs

```diff
@@ -0,0 +1,612 @@
+use pratt::Associativity;
+use pratt::PrattError;
+use pratt::PrattParser;
+use pratt::Precedence;
+
+use crate::parser::ast::BinaryOperator;
+use crate::parser::ast::Expr;
+use crate::parser::ast::Identifier;
+use crate::parser::ast::Literal;
+use crate::parser::ast::Query;
+use crate::parser::ast::TypeName;
+use crate::parser::ast::UnaryOperator;
+use crate::parser::rule::util::ident;
+use crate::parser::rule::util::literal_u64;
+use crate::parser::rule::util::Input;
+use crate::parser::rule::util::ParseError;
+use crate::parser::token::*;
+use crate::rule;
+
+pub fn query<'a, Error>(i: Input<'a>) -> IResult<Input<'a>, Query, Error>
+where Error: ParseError<Input<'a>> {
+    // TODO: unimplemented
+    nom::combinator::fail(i)
+}
+
+pub fn expr<'a, Error>(i: Input<'a>) -> IResult<Input<'a>, Expr, Error>
+where Error: ParseError<Input<'a>> {
+    subexpr(Error)(i)
+}
+
+pub fn subexpr<'a, Error>(
+    stop_at: TokenKind,
+) -> impl FnMut(Input<'a>) -> IResult<Input<'a>, Expr, Error> + Copy
+where Error: ParseError<Input<'a>> {
+    move |i| {
+        let (i, expr_elements) = rule! { (!stop_at ~ #expr_element)+ }(i)?;
+        let mut iter = expr_elements.into_iter().map(|(_, elem)| elem);
+        let expr = ExprParser
+            .parse(&mut iter)
+            .map_err(|err| match err {
+                PrattError::EmptyInput => {
+                    ContextError::add_context(i, "empty expresssion", make_error(i, ErrorKind::Eof))
+                }
+                // TODO (andylokandy): report the proper span
+                PrattError::UnexpectedNilfix(_) => ContextError::add_context(
+                    i,
+                    "unable to parse the value",
+                    make_error(i, ErrorKind::Complete),
+                ),
+                PrattError::UnexpectedPrefix(_) => ContextError::add_context(
+                    i,
+                    "unable to parse the prefix operator",
+                    make_error(i, ErrorKind::Complete),
+                ),
+                PrattError::UnexpectedInfix(_) => ContextError::add_context(
+                    i,
+                    "unable to parse the binary operator",
+                    make_error(i, ErrorKind::Complete),
+                ),
+                PrattError::UnexpectedPostfix(_) => ContextError::add_context(
+                    i,
+                    "unable to parse the postfix operator",
+                    make_error(i, ErrorKind::Complete),
+                ),
+                PrattError::UserError(_) => unreachable!(),
+            })
+            .map_err(nom::Err::Error)?;
+
+        if iter.next().is_some() {
+            return Err(nom::Err::Error(ContextError::add_context(
+                i,
+                "unable to parse the rest of expression",
+                make_error(i, ErrorKind::Complete),
+            )));
+        }
+
+        Ok((i, expr))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+#[allow(dead_code)]
+pub enum ExprElement {
```

Comment:
Would you add some comment to explain the function of this struct?

Reply:
addressed